src/storage/cache/rpc/packet.rs

```diff
@@ -0,0 +1,502 @@
+    /// The length of the request.
+    pub len: u64,
+}
+
+impl Encode for ReqHeader {
+    fn encode(&self, buf: &mut BytesMut) {
+        buf.put_u64(self.seq.to_le());
+        buf.put_u8(self.op.to_le());
+        buf.put_u64(self.len.to_le());
+    }
+}
+
+impl Decode for ReqHeader {
+    fn decode(buf: &[u8]) -> Result<Self, RpcError> {
+        if buf.len() < u64_to_usize(REQ_HEADER_SIZE) {
+            return Err(RpcError::InvalidRequest(
+                "Invalid request header".to_owned(),
+            ));
+        }
+
+        let seq = get_u64_from_buf(buf, 0)?;
+        let op = get_u8_from_buf(buf, 8)?;
+        let len = get_u64_from_buf(buf, 9)?;
+
+        Ok(ReqHeader { seq, op, len })
+    }
+}
+
+/// The message module contains the data structures shared between the client and server.
+/// Assume the cluster only contains one version server, so we won't consider the compatibility
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct RespHeader {
+    /// The sequence number of the response.
+    pub seq: u64,
+    /// The operation type of the response.
+    /// 0: keepalive
+    /// other is defined by the user
+    pub op: u8,
+    /// The length of the response.
+    pub len: u64,
+}
+
+impl Encode for RespHeader {
+    fn encode(&self, buf: &mut BytesMut) {
+        buf.put_u64(self.seq.to_le());
+        buf.put_u8(self.op);
+        buf.put_u64(self.len.to_le());
+    }
+}
+
+impl Decode for RespHeader {
+    fn decode(buf: &[u8]) -> Result<Self, RpcError> {
+        if buf.len() < u64_to_usize(RESP_HEADER_SIZE) {
+            return Err(RpcError::InvalidResponse(
+                "Invalid response header".to_owned(),
+            ));
+        }
+
+        let seq = get_u64_from_buf(buf, 0)?;
+        let op = get_u8_from_buf(buf, 8)?;
+        let len = get_u64_from_buf(buf, 9)?;
+
+        Ok(RespHeader { seq, op, len })
+    }
+}
+
+/// Define the Packet trait, for once send or receive packets
+///
+/// Client will create a packet and serialize it to a byte array,
+/// send it to the server. And create a temp response packet in client.
+///
+/// Server will receive the packet, deserialize it to a packet struct,
+/// and process the request, then create a response packet and serialize.
+///
+/// Client will receive the response packet and deserialize it to a packet struct.
+/// and check the status of the packet and the response.
+pub trait Packet: Sync + Send + Clone + Debug + Encode {
+    /// Get the packet seq number
+    fn seq(&self) -> u64;
+    /// Set the packet seq number
+    fn set_seq(&mut self, seq: u64);
+
+    /// Get packet type
+    fn op(&self) -> u8;
+    /// Set packet type
+    fn set_op(&mut self, op: u8);
+
+    /// Set timestamp
+    fn set_timestamp(&mut self, timestamp: u64);
+    /// Get timestamp
+    fn get_timestamp(&self) -> u64;
+
+    /// Get request buf length
+    /// This function is used to get the length of the request size for join header
+    fn get_req_len(&self) -> u64;
```

Comment:
Should return only the len of message, not the header,  because this len filled into Header, receiver received the fixed length header to discover how many body message pending to receive

Reply:
Yes, actually this function only return the length of the request.
This length will be filled in header's struct(length):
```rs
let req_len = req_packet.get_req_len(); // Try to get request buffer
let req_header = ReqHeader {
    seq: req_packet.seq(),
    op: req_packet.op(),
    len: req_len,
};
```